eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ExternalDependencyExposedCheck.java

```diff
@@ -0,0 +1,228 @@
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ *  No external dependency exposed in the public API
+ */
+public class ExternalDependencyExposedCheck extends AbstractCheck {
+    private static final String JAVA = "java";
+    private static final String COM_AZURE = "com.azure";
+    private static final String REACTOR = "reactor";
+    private static final String DOT = ".";
+
+    private static final String EXTERNAL_DEPENDENCY_ERROR =
+        "Class ''%s'', is a class from external dependency. You should not use it as a return or parameter argument type";
+
+    private static final Set<String> VALID_DEPENDENCY_PACKAGE_NAMES = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+        JAVA, COM_AZURE, REACTOR
+    )));
+
+    private final Map<String, String> simpleClassNameToQualifiedNameMap = new HashMap<>();
+
+    @Override
+    public void beginTree(DetailAST rootAST) {
+        simpleClassNameToQualifiedNameMap.clear();
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.IMPORT,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.IMPORT:
+                addImportedClassPath(token);
+                break;
+            case TokenTypes.METHOD_DEF:
+                checkNoExternalDependencyExposed(token);
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    /**
+     * Add all imported classes into a map, key is the name of class and value is the full package path of class.
+     *
+     * @param token the IMPORT AST node
+     */
+    private void addImportedClassPath(DetailAST token) {
+        final String importClassPath = FullIdent.createFullIdentBelow(token).getText();
+        final String className = importClassPath.substring(importClassPath.lastIndexOf(DOT) + 1);
+        simpleClassNameToQualifiedNameMap.put(className, importClassPath);
+    }
+
+    /**
+     * Checks for external dependency, log the error if it is an invalid external dependency.
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     */
+    private void checkNoExternalDependencyExposed(DetailAST methodDefToken) {
+        final DetailAST modifiersToken = methodDefToken.findFirstToken(TokenTypes.MODIFIERS);
+        if (modifiersToken == null) {
+            return;
+        }
+
+        AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
```

Comment:
I'd like to see some comments throughout your code saying what is being done, to make it easier for readers to determine. e.g. "Getting the modifier of the method to determine if it is public or protected"